configuration_go/main.go

```diff
@@ -290,6 +293,50 @@ func main() {
 		"config-w-sidecar",
 	)
 
+	// Example
+	// Thanos Compactor deployment.
+
+	// Set compactor options.
+	compactorOptions := &compactor.CompactorOptions{
+		BlockViewerGlobalSyncBlockInterval: time.Second * 5,
+		ObjstoreConfig:                     "$(OBJSTORE_CONFIG)",
+	}
+
+	// Set kube config for the compactor.
+	kubeCfg := compactor.DefaultK8sConfig()
+	kubeCfg.Replicas = 3
+	kubeCfg.Env = []corev1.EnvVar{
+		k8sutilv2.NewEnvFromSecret("OBJSTORE_CONFIG", "object_store_config", "thanos.yaml"),
+		k8sutilv2.NewEnvFromSecret("AWS_ACCESS_KEY_ID", "aws_access_key_id", "name"),
+		k8sutilv2.NewEnvFromSecret("AWS_SECRET_ACCESS_KEY", "aws_secret_access_key", "name"),
+		k8sutilv2.NewEnvFromField("HOST_IP_ADDRESS", "status.hostIP"),
+	}
+	kubeCfg.ServiceMonitor.Labels["prometheus"] = "app-sre"
+
+	// Add oauth proxy sidecar.
+	httpsPort := int32(8443)
+	oauthProxy := makeOauthProxyContainer(httpsPort, compactorOptions.HttpAddress.Port, "compactor")
+	kubeCfg.SideCars = []corev1.Container{oauthProxy}
+	kubeCfg.PodVolumes = []corev1.Volume{
+		k8sutilv2.NewPodVolumeFromSecret("compact-tls", "compact-tls"),
+		k8sutilv2.NewPodVolumeFromSecret("compact-proxy", "compact-proxy"),
+	}
+	kubeCfg.ServicePorts = []corev1.ServicePort{
+		{
+			Name:       "https",
+			Port:       httpsPort,
+			TargetPort: intstr.FromInt(int(httpsPort)),
+		},
+	}
+
+	// Create compactor, generate manifests, add annotations as post processing, generate files.
+	cpt := compactor.NewCompactor(compactorOptions, kubeCfg)
```

Comment:
With your approach, it would require shared `compactorOptions` and differentiated `kubeCfg` as opposed to a shared core struct that represents 'a compactor (and all its related bits)'

Reply:
Yes, you framed it well. Two points:

- The builder pattern used in the querier implementation looks more elegant to use but is more verbose to implement. As it is done currently, you both have to declare the `GenericConfig` with all its fields, as well as one builder per field. This doesn't provide simplification in my opinion. Directly setting the attribute of the struct seems more straightforward.
- I agree that if you want to create several deployments with a few changes:
  - The builder will be simpler to use as you just need to derive the base compactor object using the builder `With...` methods 
  - While with the factory approach here, you must clone, modify the `compactorOptions` and `kubeCfg` and finally instantiate a new compactor
  - This limitation in this PR could be circumvented by exposing these two attributes in the compactor. Then if you want to derive a base configuration, you simply modify the compactor struct. Here I made them private to limit the public part of the compactor to the minimum but did not think about this use case.
  
Should I make the `compactorOptions` and `kubeCfg` fields public in the compactor?

Reply:
I have heavily refactored the code. I think it is cleaner now. I was able to implement some kind of factory to generate kubernetes objects that should be versatile enough to be used by all thanos components. This further reduces the amount of code needed to add these components. Let me know what you think.